added uuid transform

Nearhan opened the pull request:
Added UUID transforms.
The current iteration is that you define the same structure as a concat spec, so it tries to generate UUID based on those fields. If those fields do not exist, then it will generate an UUID based on the satori/uuid library

Would love feedback in this implementation.

ryanleary commented:
At very least we need the README updated with information about the new transform type and how to use it.

I would recommend making this a bit more flexible by letting the spec specify the type of UUID (perhaps UUID v3, v4, v5), and in cases of 3 and 5 allow to specify a namespace, and finally, to be able to essentially recursively generate UUIDs (ie use base namespace plus field X for uuid, then use that as namespace + field y for the target uuid).

Nearhan commented:
@ryanleary added a ReadMe to document the spec.

Added support for v3 & v5 which recursively generates UUID for the following fields.

chilland commented on README.md:
```diff
@@ -200,8 +201,89 @@ would result in
+    "uuid": {
+      "version": 5,
+      "nameSpace": "44571214-657c-4f28-9e60-509ec39b22a9" // optional required for version 3 & 5
+
```
Should probably also add support for `require` so that behavior can be more tightly controlled when fields are missing.

Nearhan commented:
@ryanleary @chilland I've added a more designed approach would love your feedback on this.

There need to be two variations of the UUID spec one for version 4 and one for version 3 & 5.

```
{
    "operation": "uuid",
    "spec": {
        "version": 4, //required
        "uuidOutputField": "uuid" //required
    }
}
```

For version 4, its the simpliest operation. You just define what version you want to use.
Then define an `uuidOutputField` where you wish the new uuid to be placed on the doucment. 
Stragihtforward and simple.

```
{
    "operation": "uuid",
    "spec": {
        "version": 3, //required
        "nameSpace": "xxxxxxxx-xxxx-Mxxx-Nxxx-xxxxxxxxxxxx", //required
        "names": ["doc.id", "doc.auther.id"], // required
        "uuidOutputField": "doc.meta.uuid", //required
        "failOverUUID4": true //optional
    }
}
```
For versions 3 & 5, its a bit more complex. You `MUST` have a `nameSpace` field with a valid UUID.
I've decided for simplicity to have a `names` field where you an pass in paths of the document you wish to use for the names. 
It'll use the stringed value for the name recursively and the order used in the names field.
Finally it will  output the uuid in the `uuidOutputField`.
There is also a `failOverUUID4` which will allow them to just use a uuid4 in case of missing fields..

Is this flexible enough?
Is this simple enough for a users to understand?
Should we drop the failover? Might be unnecessary 

If you both give me the okay with this design I'll continue.

ryanleary commented:
That design above is far, far better. A few more change I'd like to see...

The `names` field could/should be an array of objects, which have a `path` field and a `default` field. It should pull from `path` in the document, and if it doesn't exist (or is null) use the string in `default`.  If `default` is None, then it could use the fallback behavior you described.

What is the expected behavior when `failoverUUID4` is not true or not specified? Does it just put `null` in the field?

Can you add the ability to specify some of the predefined namespaces in addition to a custom namespace.

Nearhan commented:
I think failoverUUID4 is defaulted to false. If its not there then it will be null. 
Is there a list of predefined namespaces? If so, shouldn't be too hard.

Nearhan commented:
@ryanleary @chilland 

I added the paths object and works as you mentioned. I removed the failoverUUID4 option because I think it from a user perspective, if that person wanted to use v3 or v5 and it DID fail, it would be a shock to see a UUIDv4 where you expected V5. Thus I think its simpler to let the transform fail and let the user know.

Need another look!

Nearhan commented on transform/uuid_test.go:
```diff
@@ -0,0 +1,118 @@
+
+	cfg := getConfig(spec, false)
+	kazaamOut, _ := getTransformTestWrapper(UUID, cfg, jsonIn)
+	out, _, _, _ := jsonparser.Get([]byte(kazaamOut), "a", "uuid")
```
I've done for UUIDv3 and UUIDv5 since I can know those exact UUID's for 4 i won't know the output.

chilland commented on README.md:
````diff
@@ -200,8 +202,95 @@ would result in
 ```
 
+For UUIDv3 & UUIDV5 are a bit more complex. These require a Name Space which is a valid UUID already, and a set of paths, which generate UUID's based on the value of that path. If that path doesn't exist in the incoming document, a default field will be used instead. **Note** both of these fields must be strings. 
+**Addtionally** you can use the 4 predefined namespaces such as `DNS`, `URL`, `OID`, & `X500` in the name space field otherwise pass your own UUID. 
````
sp `Additionally`

chilland commented on README.md:
````diff
@@ -200,8 +202,95 @@ would result in
+For UUIDv3 & UUIDV5 are a bit more complex. These require a Name Space which is a valid UUID already, and a set of paths, which generate UUID's based on the value of that path. If that path doesn't exist in the incoming document, a default field will be used instead. **Note** both of these fields must be strings. 
+**Addtionally** you can use the 4 predefined namespaces such as `DNS`, `URL`, `OID`, & `X500` in the name space field otherwise pass your own UUID. 
+```javascript
+{
````
For consistency, what do you think about making the spec something like this:
```javascript
 {
     "operation": "uuid",
     "spec": {
         "doc.uuid": {
           "version": 3,
           "nameSpace": "OID", 
           "names": [
             {"path": "doc.author_name", "default": "some string"},
             {"path": "doc.type", "default": "another string"}
           ]
       },
         "doc.another_uuid": {
           "version": 5,
           "nameSpace": "DNS", 
           "names": [
             {"path": "doc.author_name", "default": "some string"},
             {"path": "doc.document_id", "default": "another string"}
           ]
       }
    }
 }
 ```

Nearhan replied:
oh yeah, that is more consistent with the current api
I'll make those changes.

JoshKCarroll commented on transform/uuid.go:
```diff
@@ -0,0 +1,119 @@
+			name, _, _, err := jsonparser.Get(data, strings.Split(p, ".")...)
+			if err == jsonparser.KeyPathNotFoundError {
+
+				p, ok := field.(map[string]interface{})["default"].(string)
```
I got confused for a minute about `p` being reused here

JoshKCarroll commented on transform/uuid.go:
```diff
@@ -0,0 +1,119 @@
+		for _, field := range nameFields {
+			p, _ := field.(map[string]interface{})["path"].(string)
+
+			name, _, _, err := jsonparser.Get(data, strings.Split(p, ".")...)
```
Probably want to use `name, err := getJSONRaw(data, p, false)`, what you have here will not work properly for paths containing array indices.

Nearhan replied:
ahh.. okay

JoshKCarroll commented:
Did not look deeply at the UUID implementation, but the kazaam/json modification/general alignment with other transforms all LGTM besides couple comments I left.

Nearhan commented:
@chilland  @JoshKCarroll @ryanleary ready for another look

JoshKCarroll commented:
LGTM

chilland commented on README.md:
````diff
@@ -240,6 +241,96 @@ would result in
+```javascript
+{
+    "operation": "uuid",
+    "spec": {
````
I think the spec example needs to be updated here.

Nearhan replied:
yup just pushed it up!

Nearhan commented:
@chilland fixed up the readme
